homeassistant/components/group/entity.py

```diff
@@ -425,10 +430,34 @@ def _async_update_group_state(self, tr_state: State | None = None) -> None:
         elif tr_state.attributes.get(ATTR_ASSUMED_STATE):
             self._assumed_state = True
 
+        registry: GroupIntegrationRegistry = self.hass.data[REG_KEY]
+
+        def _update_active_on_off_states() -> tuple[set[str], set[str]]:
+            active_on_states: set[str] = set()
+            active_off_states: set[str] = set()
+            for entity_id in self.tracking:
+                if (state := self.hass.states.get(entity_id)) is not None:
+                    value = state.state
+                    domain = split_entity_id(entity_id)[0]
+                    if (
+                        domain in registry.on_states_by_domain
+                        and value in registry.on_states_by_domain[domain]
+                    ):
+                        active_on_states.add(value)
+                        # If we have more than on state, the group state
+                        # will result in STATE_ON and we can stop checking
+                        if len(active_on_states) > 1:
+                            break
+                        continue
+                    if value in registry.off_on_mapping:
+                        active_off_states.add(value)
+            return active_on_states, active_off_states
+
         num_on_states = len(self._on_states)
         # If all the entity domains we are tracking
         # have the same on state we use this state
         # and its hass.data[REG_KEY].on_off_mapping to off
+        active_on_states, active_off_states = _update_active_on_off_states()
```

Comment:
It seems very misleading and unnecessary to call `_update_active_on_off_states` right after this comment.?Instead this should be called after the `num_on_states == 1` and `num_on_states == 0` below, right?

Reply:
It needs to go before the if-block unless rewritten